src/commands/cmd_list.cc

```diff
@@ -154,6 +154,79 @@ class CommandRPop : public CommandPop {
   CommandRPop() : CommandPop(false) {}
 };
 
+class CommandLMPop : public Commander {
+ public:
+  /// format: LMPOP #numkeys key0 [key1 ...] <LEFT | RIGHT> [COUNT count]
+  Status Parse(const std::vector<std::string> &args) override {
+    CommandParser parser(args, 1);
+
+    auto num_keys = GET_OR_RET(parser.TakeInt<uint32_t>());
+    keys_.clear();
+    keys_.reserve(num_keys);
+    for (uint32_t i = 0; i < num_keys; ++i) {
+      keys_.emplace_back(GET_OR_RET(parser.TakeStr()));
+    }
+
+    auto left_or_right = util::ToLower(GET_OR_RET(parser.TakeStr()));
+    if (left_or_right == "left") {
+      left_ = true;
+    } else if (left_or_right == "right") {
+      left_ = false;
+    } else {
+      return {Status::RedisParseErr, errInvalidSyntax};
+    }
+
+    if (parser.Good()) {
+      if (parser.EatEqICase("count")) {
+        count_ = GET_OR_RET(parser.TakeInt<uint32_t>());
+      } else
+        return parser.InvalidSyntax();
+    }
+
+    return parser.Good() ? parser.InvalidSyntax() : Status::OK();
+  }
+
+  Status Execute(Server *svr, Connection *conn, std::string *output) override {
+    redis::List list_db(svr->storage, conn->GetNamespace());
+
+    std::vector<std::string> elems;
+    std::string chosen_key;
+    for (auto &key : keys_) {
+      auto s = list_db.PopMulti(key, left_, count_, &elems);
+      if (!s.ok() && !s.IsNotFound()) {
+        return {Status::RedisExecErr, s.ToString()};
+      }
+      if (!elems.empty()) {
+        chosen_key = key;
+        break;
+      }
+    }
+
+    if (elems.empty()) {
+      *output = redis::NilString();
+    } else {
+      std::string elems_bulk = redis::MultiBulkString(elems);
+      *output = redis::Array({redis::BulkString(chosen_key), elems_bulk});
+    }
+
+    return Status::OK();
+  }
+
+  static const inline CommandKeyRangeGen keyRangeGen = [](const std::vector<std::string> &args) {
+    CommandKeyRange range;
+    range.first_key = 2;
+    range.key_step = 1;
+    auto v = ParseInt<int32_t>(args[1], 10);
```

Comment:
Guess we should check the range of `v` here?

Reply:
There are two reasons I didn't do the checking here. 

Firstly, in the call chain, the command would be parsed before this keyRangeGen is invoked, so if the range is invalid then the parsing phase would report error. 

Secondly (and most importantly...), I am not aware how to handle the exception/error if there is something wrong here. The error handling style of this project is mostly via the returned error code, and the rangeGen interface has no such error return value. I am not sure what to do if there is an error here. Should I throw some exception?

Reply:
ok, it makes sense, thank you.

Reply:
Yeah, the parsing will be invoked before this method, so technically the ParseInt will always succeed.